modules/transaction-payment/src/lib.rs

```diff
@@ -611,60 +673,198 @@ where
-				match trading_path.last() {
-					Some(target_currency_id) if *target_currency_id == native_currency_id => {
-						let supply_currency_id = *trading_path.first().expect("these's first guaranteed by match");
-						// calculate the supply limit according to oracle price and the slippage limit,
-						// if oracle price is not avalible, do not limit
-						let max_supply_limit = if let Some(target_price) =
-							T::PriceSource::get_relative_price(*target_currency_id, supply_currency_id)
-						{
-							Ratio::one()
-								.saturating_sub(T::MaxSwapSlippageCompareToOracle::get())
-								.reciprocal()
-								.unwrap_or_else(Ratio::max_value)
-								.saturating_mul_int(target_price.saturating_mul_int(amount))
-						} else {
-							PalletBalanceOf::<T>::max_value()
-						};
+		// make sure add extra gap to keep alive after swap.
+		let amount = fee.saturating_add(native_existential_deposit.saturating_sub(total_native));
+		// native is not enough, try swap native from treasury fee pool to pay fee and gap.
+		Self::swap_native_asset(who, amount);
+	}
 
-						if T::DEX::swap_with_exact_target(
-							who,
-							&trading_path,
-							amount.unique_saturated_into(),
-							<T as Config>::MultiCurrency::free_balance(supply_currency_id, who)
-								.min(max_supply_limit.unique_saturated_into()),
-						)
-						.is_ok()
-						{
-							// successfully swap, break iteration
-							break;
-						}
+	/// Iterate order list, if can swap out enough native asset amount with foreign asset, then
+	/// stop.
+	fn swap_native_asset(who: &T::AccountId, amount: Balance) {
+		let native_currency_id = T::NativeCurrencyId::get();
+		for trading_path in Self::get_trading_path(who) {
+			if let Some(target_currency_id) = trading_path.last() {
+				if *target_currency_id == native_currency_id {
+					let supply_currency_id = *trading_path.first().expect("should match a non native asset");
+					if Self::swap_from_treasury(who, amount, supply_currency_id).is_ok() {
+						break;
 					}
-					_ => {}
 				}
 			}
 		}
 	}
+
+	/// swap from treasury fee pool.
+	#[transactional]
+	fn swap_from_treasury(
+		who: &T::AccountId,
+		amount: Balance,
+		supply_currency_id: CurrencyId,
+	) -> sp_std::result::Result<(), DispatchError> {
+		let treasury_account = Self::treasury_sub_account_id(supply_currency_id);
+		let treasury_supply_balance = T::MultiCurrency::free_balance(supply_currency_id, &treasury_account);
+		let mut rate = TokenFixedRate::<T>::get(supply_currency_id).expect("supply asset not in pool.");
+
+		// user's supply asset should have enough balance based on old fix rate.
+		ensure!(
+			T::MultiCurrency::free_balance(supply_currency_id, who) >= rate.saturating_mul_int(amount),
+			Error::<T>::SupplyBalanceTooLow
+		);
+
+		// if treasury account has not enough native asset, trigger swap
+		let treasury_balance = T::Currency::free_balance(&treasury_account);
+		if treasury_balance < SwapBalanceThreshold::<T>::get(supply_currency_id) {
```

Comment:
`treasury_supply_balance` maybe is 0.
It seems we don’t need `SwapBalanceThreshold`. We need to swap all of the tokens(or the largest value) in the `on_initialize`.
@xlc How do you think about this?

Reply:
add some doc. we normaly set threshold >= ED to prevent exhausted treasury balance. but there maybe cases that treasury supply balance is <= ED(but it's not 0, because the treasury sub account is in our whitelist), it that happens, the swap may failed.